File: carmap_with_edgewarp/CarMap_Cloud_Service/store/StateTracker.cpp
#include "StateTracker.h"
#include <memory>
#include <chrono>
#include <limits>

StateTracker::StateTracker(shared_ptr<mutex> m_tx)
{
    StateTracker::AppStartTime = std::chrono::steady_clock::now();
    mtx = m_tx;
}

StateTracker::~StateTracker()
{
}

void StateTracker::AddKey(string key, uint32_t size)
{
    shared_ptr<StateInfo> info = make_shared<StateInfo>(key);
    info->keySize = size;
    StateTracker::StateStore.insert(std::pair<string, shared_ptr<StateInfo>>(key, info));
}

void StateTracker::UpdateKey(string key)
{
    // If key exists
    if (StateStore.count(key) > 0)
    {
        StateStore[key]->updateTime = std::chrono::steady_clock::now();
    }
    else
    {
        shared_ptr<StateInfo> info = make_shared<StateInfo>(key);
        info->keySize = 0;
        StateTracker::StateStore.insert(std::pair<string, shared_ptr<StateInfo>>(key, info));
    }

    StateStore[key]->updateCounter = StateStore[key]->updateCounter + 1;
}

void StateTracker::UpdateSyncTime(string key, std::chrono::steady_clock::time_point time)
{
    if (StateStore.count(key) > 0)
    {
        StateStore[key]->syncTime = time;
        StateStore[key]->synCounter = StateStore[key]->synCounter + 1;
    }
}

vector<string> StateTracker::GetOldestUpdates()
{
    vector<string> keys;

    // Latest time.
    auto tempTime = std::chrono::steady_clock::now();

    for (auto item : StateStore)
    {
        if (item.second->keySize > KeySizeUpperLimit)
        {
            if (tempTime > item.second->updateTime && item.second->syncTime < item.second->updateTime)
            {
                tempTime = item.second->updateTime;
                keys.push_back(item.first);
            }
        }
        else
        {
            if (item.second->GetFrequency() <= UpdateRateUpperLimit)
            {
                if (tempTime > item.second->updateTime && item.second->syncTime < item.second->updateTime)
                {
                    tempTime = item.second->updateTime;
                    keys.push_back(item.first);
                }
            }
        }
    }
    

    return keys;
}

vector<string> StateTracker::GetOutOfSyncKeys()
{
    vector<string> keys;

    for (auto item : StateStore)
    {
        if (item.second->updateTime > item.second->syncTime)
        {
            keys.push_back(item.first);
        }
    }

    return keys;
}

std::chrono::steady_clock::time_point StateTracker::GetLatestUpdateTime()
{
    auto tempItem = StateStore.at(0);

    for (auto item : StateStore)
    {
        if (tempItem->updateTime < item.second->updateTime)
        {
            tempItem->updateTime = item.second->updateTime;
        }
    }

    return tempItem->updateTime;
}

void StateTracker::PrintKeys()
{
    for (auto item : StateStore)
    {
        cout << "Key = " << item.first << ", Size = " << item.second->keySize << ", Update Counter = " << item.second->updateCounter << ", Update Frequency = " << item.second->updateFrequency << ", Sync Counter = " << item.second->synCounter << endl;
    }
}
// void StateTracker::AllSyncDone();

File: emp_with_edgewarp/src/main/java/org/emp/store/StateTracker.java
package org.emp.store;

import lombok.AccessLevel;
import lombok.Builder;
import lombok.Data;
import lombok.experimental.FieldDefaults;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Data
@FieldDefaults(level = AccessLevel.PRIVATE)
public class StateTracker {
    final float UpdateRateUpperLimit = 10; // todo: make it configurable
    // Don't sync keys bigger than 10 KBs in advance.
    final long KeySizeUpperLimit = 10*1024; // todo: make it configurable
    long AppStartTime;

    Map<String, StateInfo> StateStore = new HashMap<String, StateInfo>();

    StateTracker()
    {
        AppStartTime = System.currentTimeMillis();
    }

    public void AddKey(String key, long size)
    {
        StateInfo info = new StateInfo(key);
        info.setKeySize(size);
        StateStore.put(key, info);
    }

    public void UpdateKey(String key)
    {
        // If key exists
        if (StateStore.containsKey(key))
        {
            StateStore.get(key).setUpdateTime(System.currentTimeMillis());
        }
        else
        {
            StateInfo info = new StateInfo(key);
            info.setKeySize(0);
            StateStore.put(key, info);
        }
        StateStore.get(key).setUpdateCounter(StateStore.get(key).getUpdateCounter() + 1);
    }

    public void UpdateSyncTime(String key, long time)
    {
        if (StateStore.containsKey(key))
        {
            StateStore.get(key).setSyncTime(time);
            StateStore.get(key).setSynCounter(StateStore.get(key).getSynCounter() + 1);
        }
    }

    public List<String> GetOldestUpdates()
    {
        List<String> keys = new ArrayList<String>();

        // Latest time.
        long tempTime = System.currentTimeMillis();

        for (Map.Entry<String,StateInfo> item : StateStore.entrySet())
        {
            String key = item.getKey();
            StateInfo value = item.getValue();
            if (value.getKeySize() > KeySizeUpperLimit)
            {
                if (tempTime > value.getUpdateTime() && value.getSyncTime() < value.getUpdateTime()) {
                    tempTime = value.getUpdateTime();
                    keys.add(key);
                }
            } else
            {
                if (value.GetFrequency() <= UpdateRateUpperLimit) {
                    if (tempTime > value.getUpdateTime() && value.getSyncTime() < value.getUpdateTime()) {
                        tempTime = value.getUpdateTime();
                        keys.add(key);
                    }
                }
            }
        }


        return keys;
    }

    public List<String> GetOutOfSyncKeys()
    {
        List<String> keys = new ArrayList<>();

        for (Map.Entry<String,StateInfo> item : StateStore.entrySet())
        {
            StateInfo value = item.getValue();
            if (value.getUpdateTime() > value.getSyncTime())
            {
                keys.add(item.getKey());
            }
        }

        return keys;
    }

    public long GetLatestUpdateTime()
    {
        long time = 0;

        for (Map.Entry<String,StateInfo> item : StateStore.entrySet())
        {
            if (time < item.getValue().getUpdateTime())
            {
                time = item.getValue().getUpdateTime();
            }
        }

        return time;
    }

    public void PrintKeys()
    {
        for (Map.Entry<String,StateInfo> item : StateStore.entrySet())
        {
            StateInfo value = item.getValue();
            System.out.println("Key = " + item.getKey() + ", Size = " + value.getKeySize() + ", Update Counter = " + value.getUpdateCounter() + ", Update Frequency = " + value.getUpdateFrequency() + ", Sync Counter = " + value.getSynCounter());
        }
    }
}
